routing/turns_generator.cpp

```diff
@@ -806,5 +893,70 @@ void GetTurnDirection(Index const & index, TurnInfo & turnInfo, TurnItem & turn)
     return;
   }
 }
+
+size_t CheckUTurnOnRoute(vector<LoadedPathSegment> const & segments, size_t currentSegment, TurnItem & turn)
+{
+  size_t constexpr kUTurnLookAhead = 3;
+  double constexpr kUTurnHeadingSensitivity = math::pi / 10.0;
+
+  // In this function we process the turn between the previous and the current
+  // segments. So we need a shift to get the previous segment.
+  ASSERT_GREATER(segments.size(), 1, ());
+  ASSERT_GREATER(currentSegment, 0, ());
+  ASSERT_GREATER(segments.size(), currentSegment, ());
+  auto const & masterSegment = segments[currentSegment - 1];
+  ASSERT_GREATER(masterSegment.m_path.size(), 1, ());
+  // Roundabout is not the UTurn.
+  if (masterSegment.m_onRoundabout)
+    return 0;
+  for (size_t i = 0; i < min(kUTurnLookAhead, segments.size() - currentSegment); ++i)
```

Comment:
To prevent possible crashes in release, it's still better to write loop condition like:
```c++
i < kUTurnLookAhead && i + currentSegment < segments.size()
```
I see an assertion above, but it will be thrown away in release build.